### PR TITLE
Free text search

### PR DESCRIPTION
**Pull request by eldang**
This gets us 2/3 of the way on #13 and I think it's probably wise for me to stop here.  It adds a free text entry box for zooming to ISDs, which responds as soon as enough of an ISD's name has been typed to get a unique match.

It does not yet present the user with a set of autocomplete options, but I'm not so sure that's important now having played with this a bit.

I've tried to write this in such a way that if there's a later request for an ESCs free text search it'll be easy to adapt.

The input needs styling.

The code changes here should be orthogonal enough from #28 that either can be merged independently of the other.

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -90,6 +90,8 @@
 
 			<!--Drop down controls-->
 
+			<input id="school-districts-text-entry" name="school-districts-text-entry" type="text" placeholder="enter ISD name" oninput="textZoomHandler('isd', 'state-school-districts', 'NAME', this.value);" />
```
Comment (coregis) [inline index.html:93]: I'm sure most people using this map will know what ISD stands for, but it might be more intuitive to say "enter District name" here

**Comment by coregis** [comment]
> It does not yet present the user with a set of autocomplete options, but I'm not so sure that's important now having played with this a bit.

This free text entry feature is really cool! It does feel a little black-boxy to me, especially the way it zooms-to-district as soon as there is only one option that fits the free text. I also think it might be possible that a person knows only a portion of the district name they're searching for, and without any suggestions/options they might just be stuck with a partial name and no guidance about what to do next. So, how much effort is it to add this functionality? Specifically, I'm wondering if you are able to make the free text entry box do 2 things:

1. Generate a list of options
2. Require the user to click on a selection before zooming, so there is no ambiguity about where and when the map will take them there

**Comment by eldang** [comment]
@coregis this makes sense to me.  I was talking to @nein09 about this and we think it'll make sense for her to do most of the next part, since it's mostly UI work and the quality of that will be what determines how much anyone uses this feature.

@nein09 some notes for when you pick this up:

- I can definitely help with the workings, but I don't think you're going to have trouble understanding the new code.
- The existing code it leans on is a function written for the drop-down menus, which expects to be given a bounding box as a zoom target, and the drop-downs themselves store those bounding boxes.  So that's why I've added a data structure to allow looking up bounding boxes by district name, which wasn't previously needed, and gets populated at the same time that the drop-down menus are.
- I think it'll probably make sense to cut `findDistrictBySubstring()` out if we're showing people autocomplete options, and show those options as soon as there are a manageable number.  I think for most district names that'll be the case by the time someone's typed 3 or 4 letters.
- I _think_ the reason I used title case for the reference values rather than just lowercasing everything is that the district name also ends up being used for things like the stats box and page title.

**Comment by nein09** [comment]
I found a goofy little thing that pretty much does what we want: https://tom-select.js.org/. If you're okay with using a third-party solution, and you like how it looks, I think I can do a little cleanup in here, and call this good.

**Comment by coregis** [comment]
> I found a goofy little thing that pretty much does what we want: https://tom-select.js.org/. If you're okay with using a third-party solution, and you like how it looks, I think I can do a little cleanup in here, and call this good.

This is working really well and I am 100% fine with using a third-party solution! I think all we have left is to switch from ESC City to Region Number then we can stick a fork in this

**Comment by eldang** [comment]
This is working nicely, so I'm going to resolve the merge conflict and then merge it, before starting #31 .

One thought though: should we give the user any more of a prompt that they can type in the School Districts box?  I think it's easy to miss that it's anything other than a dropdown menu.

**Comment by eldang** [comment]
OK, merge conflicts sorted.

File: scripts/functions.js [hunk 2]
```diff
@@ -703,15 +703,10 @@ function updateStatsBox() {
...
-			if (loadedPointLayers[i][0].includes("raising-blended-learners")) {
-				f = ["<", "year", (filterStates.year + 4).toString()];
-			} else {
-				f = ["==", "year", filterStates.year.toString()];
-			}
 			pointsInDistrict = getUniqueFeatures(
 				map.queryRenderedFeatures({
 					layers: [loadedPointLayers[i][0]],
-					filter: f,
+					filter: ['<=', 'year', filterStates.year.toString()],
```
Comment (nein09) [inline scripts/functions.js:709]: my apologies!

File: scripts/functions.js [hunk 3]
```diff
@@ -230,54 +230,6 @@ function populateZoomControl(
...
-				sourceID,
-				districtBBOXes[districtType][districtName].toString() +
-					"," +
-					districtName,
-				fieldName,
-			);
-		}
-	}
-}
-
```
Comment (eldang) [inline scripts/functions.js:182]: ✂️✂️✂️

**Pull request merged**